rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -2033,11 +2033,14 @@ extern "C" rmw_publisher_t * rmw_create_publisher(
 
 extern "C" rmw_ret_t rmw_get_gid_for_publisher(const rmw_publisher_t * publisher, rmw_gid_t * gid)
 {
-  RET_NULL(publisher);
-  RET_WRONG_IMPLID(publisher);
-  RET_NULL(gid);
+  RMW_CHECK_ARGUMENT_FOR_NULL(publisher, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    publisher,
+    publisher->implementation_identifier,
+    eclipse_cyclonedds_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RMW_CHECK_ARGUMENT_FOR_NULL(gid, RMW_RET_INVALID_ARGUMENT);
   auto pub = static_cast<const CddsPublisher *>(publisher->data);
-  RET_NULL(pub);
```

Comment:
So, you don't need to check if publisher->data was casteable?

Reply:
To `static_cast` a `void *` gives no guarantees as to whether that pointer actually refers to such object type e.g. `publisher->data` could be equal to `0xFFFFFFFF`, and it'd still succeed. 

To check for nullity is insufficient at best, harmful at worst. Insufficient because it's the canonical invalid pointer but not the only one -- and we cannot detect any other invalid pointer. Harmful because it deals with the situation as if it was an expected runtime error. It's not. If `publisher->data` is invalid but the publisher belongs to this implementation, that means that either the caller broke API preconditions, the implementation itself has logical errors, or memory has been corrupted. In any case, it's UB.

Reply:
So, checking if `publisher->data` is not null makes no sense either?

Reply:
Correct.

Reply:
Same as https://github.com/ros2/rmw_cyclonedds/pull/244#discussion_r492741744